### ISSUE TITLE
docker-compose build ignoring .dockerignore?

### ISSUE THREAD

**Issue by devinrsmith** [docker/compose#1607]
I've got a full example repo here:

https://github.com/devinrsmith/docker-compose-build-test

If I'm using `docker-compose build` incorrectly, please let me know!

**Comment by neciu** [comment]
You have problem with format.

In your `.dockerignore` change `subdir/` to `subdir`.

**Comment by devinrsmith** [comment]
Why does the .dockerignore work for `docker build -t subdir ../` but not `docker-compose build`?

**Comment by fChristenson** [comment]
I am having the same issue running docker-compose 1.2.0.
Running vanilla docker build pulls in the ignore file and builds the image, np there.
However docker-compose does not use the ignore file, with or without the trailing slash.

**Comment by aanand** [comment]
Looks like a docker-py issue; a fix is being worked on in https://github.com/docker/docker-py/pull/604. Marking this as a bug to remind us to update our docker-py version once the fix is out.

**Comment by thaJeztah** [comment]
@aanand looks like https://github.com/docker/docker-py/pull/604 stalled and was removed from the milestone there?

**Comment by MrMMorris** [comment]
my `.dockerignore` is:
`*.sql`

I have `2015-07-22.sql` in my build dir

and yet:
`docker exec -it container ls | grep *.sql`
`2015-07-22.sql`

**Comment by aanand** [comment]
@thaJeztah I've created https://github.com/docker/docker-py/pull/721 to continue the work.

**Comment by twillouer** [comment]
Hi,

I'm facing one strange problem with .dockerignore:

```
*.sh

!awesome/script.sh
```

with docker build all is fine, but docker-compose didn't see "awesome/script.sh"

Am I in the same issue ?

**Comment by aanand** [comment]
@twillouer that's a known bug, fixed by https://github.com/docker/docker-py/pull/721 - it'll be fixed on the Compose side in 1.5.0.

**Comment by twillouer** [comment]
thanks !

**Comment by JonasT** [comment]
This still seems to be broken even on latest git master on some machines. It doesn't work for me on a machine given this configuration:

Debian 8.2 with kernel 3.16.0 with lxc-docker version 1.7.1, build 786b29d
docker-compose is git master, HEAD is at dabf1e8657674014a5bc89f99edbf2fe0629bb71

The .dockerignore works fine for docker build (which runs instantly), but not for docker-compose (which uploads stuff it shouldn't upload for various minutes).

Also see #2100

**Comment by cruelbob** [comment]
I have same problem. docker-compose ignores .dockerignore. "docker build" works fine.
System: windows 10

**Comment by xrstf** [comment]
I'm experiencing the same problem using the Docker Toolbox that was released yesterday:

* Windows 7 x64
* Docker version 1.9.0, build 76d6bc9
* Docker-Compose 1.5.0

**Comment by dnephin** [comment]
It sounds like we're still having problems with `.dockerignore`. If you're experiencing these problems with compose 1.5.0, please include a sample `.dockerignore` and directory structure so we can reproduce the error.

**Comment by xrstf** [comment]
Sorry to not provide a smaller usecase, but this is how I noticed the problem:

I have this ``package.json`` for my project:

```json
{
  "dependencies": {
    "grunt-contrib-uglify": "^0.9.2"
  }
}
```

Plus I have this ``docker-compose.yml``:

```yaml
web:
  build: .
```

The ``Dockerfile`` looks like this:

```
FROM node:0.12
```

My ``.dockerignore`` contains one line (I tried adding a trailing slash, but the problem persisted):

```
node_modules
```

Now I can do

```
> npm install
  (...snip...)
> docker build .
Sending build context to Docker daemon  5.12 kB
```

Awesome, it works, only 5 KiB are sent (``node_modules`` is roughly 10 MiB in total). With ``docker-compose`` in ``Q:\sites\test``:

```
> npm install
  (...snip...)
> docker-compose up
Building web
Traceback (most recent call last):
  File "D:\opt\python\Scripts\docker-compose-script.py", line 9, in <module>
    load_entry_point('docker-compose==1.5.0dev', 'console_scripts', 'docker-compose')()
  File "D:\opt\python\lib\site-packages\compose\cli\main.py", line 54, in main
    command.sys_dispatch()
  File "D:\opt\python\lib\site-packages\compose\cli\docopt_command.py", line 23, in sys_dispatch
    self.dispatch(sys.argv[1:], None)
  File "D:\opt\python\lib\site-packages\compose\cli\docopt_command.py", line 26, in dispatch
    self.perform_command(*self.parse(argv, global_options))
  File "D:\opt\python\lib\site-packages\compose\cli\main.py", line 170, in perform_command
    handler(project, command_options)
  File "D:\opt\python\lib\site-packages\compose\cli\main.py", line 583, in up
    detached=detached
  File "D:\opt\python\lib\site-packages\compose\project.py", line 313, in up
    detached=detached
  File "D:\opt\python\lib\site-packages\compose\service.py", line 404, in execute_convergence_plan
    container = self.create_container(do_build=do_build)
  File "D:\opt\python\lib\site-packages\compose\service.py", line 303, in create_container
    self.ensure_image_exists(do_build=do_build)
  File "D:\opt\python\lib\site-packages\compose\service.py", line 326, in ensure_image_exists
    self.build()
  File "D:\opt\python\lib\site-packages\compose\service.py", line 718, in build
    dockerfile=self.options.get('dockerfile', None),
  File "D:\opt\python\lib\site-packages\docker\api\build.py", line 48, in build
    context = utils.tar(path, exclude=exclude, dockerfile=dockerfile)
  File "D:\opt\python\lib\site-packages\docker\utils\utils.py", line 85, in tar
    t.add(os.path.join(root, path), arcname=path, recursive=False)
  File "D:\opt\python\lib\tarfile.py", line 1998, in add
    tarinfo = self.gettarinfo(name, arcname)
  File "D:\opt\python\lib\tarfile.py", line 1870, in gettarinfo
    statres = os.lstat(name)
WindowsError: [Error 3] Das System kann den angegebenen Pfad nicht finden: 'Q:\\sites\\test\\node_modules\\grunt-contrib-uglify\\node_modules\\maxmin\\node_modules\\pretty-bytes\\node_modules\\meow\\node_modules\\normalize-package-data\\node_modules\\validate-npm-package-license\\node_modules\\spdx-correct\\node_modules\\spdx-license-ids\\spdx-license-ids.json'
```

(I am using 1.5.0dev on this machine, but I got the exact same problem on my other machine with 1.5.0 final).

With node, I often encounter problems with too long paths and stuff on Windows, but the fact that Docker-Compose is attempting to tar up the node_modules is the indicator that the ``.dockerignore`` is ignored.

**Comment by JonasT** [comment]
FWIW, I am still having issues with this on Linux (Ubuntu) with docker 1.5.0dev, so it's probably not isolated to Windows. However, it's on a production machine so I cannot easily put together a minimal test case (it works fine on my testing machine).

**Comment by Quantumplation** [comment]
I am also having this issue.

Directory structure:
```
docker-compose.yml
web
  + .dockerignore
  + Dockerfile
  + node_modules
    + ...
```

docker-compose.yaml
```
web:
  build: web
  tty: true
  ports:
    - 8081:5000
```

Dockerfile
```
FROM microsoft/aspnet

# Curl, node, npm, bower, grunt
RUN apt-get update && apt-get install -y curl
RUN curl -sL https://deb.nodesource.com/setup | bash -
RUN apt-get install -y nodejs
RUN npm install -g bower
RUN npm install -g grunt-bower-cli
RUN npm install -g grunt
RUN npm install -g grunt-cli
RUN npm install -g grunt-bower-task

# Copy the project.json file first, then do a restore.
# This ensures that as long as project.json doesn't change, it will avoid
# doing a package restore
COPY project.json /app/
COPY bower.json /app/
COPY gruntfile.js /app/
COPY package.json /app/
WORKDIR /app

RUN ["dnu", "restore"]

# Then copy the rest of the files
COPY . /app

# Expose the port that the website listens on
EXPOSE 5000
# And start the website
ENTRYPOINT ["dnx", "-p", "project.json", "web"]
```

.dockerignore
```
node_modules
```


Result:

```
Pi@Ricci MINGW64 /d/proj/Repro
$ docker-compose build
Building web
Traceback (most recent call last):
  File "<string>", line 3, in <module>
  File "C:\projects\compose\compose\cli\main.py", line 54, in main
  File "C:\projects\compose\compose\cli\docopt_command.py", line 23, in sys_dispatch
  File "C:\projects\compose\compose\cli\docopt_command.py", line 26, in dispatch
  File "C:\projects\compose\compose\cli\main.py", line 171, in perform_command
  File "C:\projects\compose\compose\cli\main.py", line 192, in build
  File "C:\projects\compose\compose\project.py", line 235, in build
  File "C:\projects\compose\compose\service.py", line 683, in build
  File "c:\projects\compose\venv\lib\site-packages\docker\api\build.py", line 48, in build
  File "c:\projects\compose\venv\lib\site-packages\docker\utils\utils.py", line 85, in tar
  File "c:\python27-x64\Lib\tarfile.py", line 2000, in add
  File "c:\python27-x64\Lib\tarfile.py", line 1872, in gettarinfo
WindowsError: [Error 3] The system cannot find the path specified: 'D:\\proj\\Repro\\web\\node_modules\\babel-preset-react\\node_modules\\babel-plugin-transform-react-jsx\\node_modules\\babel-helper-builder-react-jsx\\node_modules\\babel-types\\node_modules\\babel-traverse\\node_modules\\babel-code-frame\\node_modules\\js-tokens\\changelog.md'
docker-compose returned -1
```

**Comment by keatz55** [comment]
+1

**Comment by keatz55** [comment]
Did anybody find a workaround?  I am trying to do everything in only the Dockerfile, but haven't had any luck yet.

**Comment by esc-rtn** [comment]
I'm getting this issue as well, could it be related to the node version everyone is running? Via this issue: https://github.com/npm/npm/issues/3697

**Comment by esc-rtn** [comment]
+1

**Comment by JonasT** [comment]
Again despite this being apparently ignored, I am seeing this on a Linux machine as well.

**Comment by esc-rtn** [comment]
I fixed this issue by removing an unnecessary volume reference in my docker-compose file. Compose ignores .gitignore for volumes. Just FYI

**Comment by thaJeztah** [comment]
@esc-rtn the `.dockerignore` file only specifies what files should not be sent to the daemon during *build*. If you're using a bind-mounted volume during "run", it will simply mount all files in that location as a volume (all files that are present on the host).

**Comment by stekershaw** [comment]
I'm seeing similar too, on Windows: `docker build` works as expected, `docker-compose` not.

I've put a small example on GitHub, with notes in the README.md: https://github.com/stekershaw/docker-compose-ignore

**Comment by dnephin** [comment]
Another round of fixes for `.dockerignore` went into docker-py after Compsoe 1.5.2 was released. Could you try out the Compose 1.6.0 RC2 release to see if it's fixed now?

**Comment by stekershaw** [comment]
Just tried with docker-compose version 1.6.0rc2, build a7636be.

I have three test cases in my repo and with 1.5.2 two of them failed. Now with 1.6.0rc2 only one fails, when I have a directory and a file in the directory excluded, like so:
```
$ cat .dockerignore
files/test_dir
!files/test_dir/should_be_here_maybe
```

**Comment by aanand** [comment]
I get identical behaviour in that case between `docker build` and `docker-compose build`.

`.dockerignore`:

```
files/test_dir
!files/test_dir/should_be_here_maybe
```

`docker-compose.yml`:

```yaml
test:
  build: .
```

`Dockerfile`:

```dockerfile
FROM busybox
COPY . /context
CMD ["find", "/context"]
```

```console
$ mkdir -p files/test_dir
$ touch files/test_dir/should_be_here_maybe

$ find .
.
./.dockerignore
./docker-compose.yml
./Dockerfile
./files
./files/test_dir
./files/test_dir/should_be_here_maybe

$ docker build --no-cache -t 1607-docker-build .
Sending build context to Docker daemon  5.12 kB
Step 1 : FROM busybox
 ---> 0cb40641836c
Step 2 : COPY . /context
 ---> 859e12600100
Removing intermediate container 9067b263098b
Step 3 : CMD find /context
 ---> Running in 1ddc0a573492
 ---> b1b3beacf5f2
Removing intermediate container 1ddc0a573492
Successfully built b1b3beacf5f2

$ docker run 1607-docker-build
/context
/context/docker-compose.yml
/context/Dockerfile
/context/files
/context/files/test_dir
/context/files/test_dir/should_be_here_maybe
/context/.dockerignore

$ docker-compose build --no-cache
Building test
Step 1 : FROM busybox
 ---> 0cb40641836c
Step 2 : COPY . /context
 ---> d86507051d6d
Removing intermediate container 0af2cbf69b17
Step 3 : CMD find /context
 ---> Running in 8533dae3af74
 ---> 1f736ecb2b38
Removing intermediate container 8533dae3af74
Successfully built 1f736ecb2b38

$ docker-compose run test
/context
/context/docker-compose.yml
/context/Dockerfile
/context/files
/context/files/test_dir
/context/files/test_dir/should_be_here_maybe
/context/.dockerignore
```

**Comment by stekershaw** [comment]
Hi @aanand, thanks for you response, I get the same result as you following your steps.

However, the test that I ran previously also had another file present in files/test_dir. If I add another file to files/test_dir then I see that it is not present (as I would expect) with `docker build` but it is present with `docker-compose`:

```
$ touch files/test_dir/should_not_be_here

$ docker build --no-cache -t 1607-docker-build .
Sending build context to Docker daemon  5.12 kB
Step 1 : FROM busybox
 ---> b175bcb79023
Step 2 : COPY . /context
 ---> a23d9645c21c
Removing intermediate container 8eb2bb23c4db
Step 3 : CMD find /context
 ---> Running in d9fef847acd8
 ---> e52ae84b1250
Removing intermediate container d9fef847acd8
Successfully built e52ae84b1250
SECURITY WARNING: You are building a Docker image from Windows against a non-Windows Docker host. All files and directories added to build context will have '-rwxr-xr-x' permissions. It is recommended to double check and reset permissions for sensitive files and directories.

$ docker run --rm 1607-docker-build
/context
/context/Dockerfile
/context/files
/context/files/test_dir
/context/files/test_dir/should_be_here_maybe
/context/docker-compose.yml
/context/.dockerignore

$ docker-compose build --no-cache
Building test
Step 1 : FROM busybox
 ---> b175bcb79023
Step 2 : COPY . /context
 ---> 9df0cf4bfb69
Removing intermediate container 7820f982d59e
Step 3 : CMD find /context
 ---> Running in 06e1a0b89a45
 ---> 2c922dbc66d9
Removing intermediate container 06e1a0b89a45
Successfully built 2c922dbc66d9

$ docker-compose run test
ERROR: Interactive mode is not yet supported on Windows.
Please pass the -d flag when using `docker-compose run`.

$ docker-compose run -d test
dcitest_test_run_2

$ docker logs dcitest_test_run_2
/context
/context/Dockerfile
/context/files
/context/files/test_dir
/context/files/test_dir/should_not_be_here
/context/files/test_dir/should_be_here_maybe
/context/docker-compose.yml
/context/.dockerignore
``` 

This is Windows (using git bash as my shell), checked with both with Docker compose 1.5.2 and 1.6.0rc2, docker 1.9.1.

Running the same on Ubuntu 14.04 with docker-compose 1.5.2 and docker 1.9.1 is fine:

```
# cat .dockerignore
files/test_dir
!files/test_dir/should_be_here_maybe

# find .
.
./.dockerignore
./docker-compose.yml
./files
./files/test_dir
./files/test_dir/should_not_be_here
./files/test_dir/should_be_here_maybe
./Dockerfile

# docker-compose build --no-cache
Building test
Step 1 : FROM busybox
 ---> b175bcb79023
Step 2 : COPY . /context
 ---> c533a0768d5e
Removing intermediate container 0c057fe8eb82
Step 3 : CMD find /context
 ---> Running in e8a0cf1f58d8
 ---> 175777486a25
Removing intermediate container e8a0cf1f58d8
Successfully built 175777486a25

# docker-compose run test
/context
/context/docker-compose.yml
/context/.dockerignore
/context/files
/context/files/test_dir
/context/files/test_dir/should_be_here_maybe
/context/Dockerfile
```

**Comment by bkcummins** [comment]
Still dealing with this most of the afternoon.

Docker version 1.10.1, build 9e83765
docker-compose version 1.6.0, build d99cad6

````
#docker-compose.yml
test:
  build: ./cdn
````

```
#Dockerfile
FROM busybox
COPY ["content/","/test/"]
RUN find /test/ -maxdepth 1
```

```
#.dockerignore
**/.DS_Store
**/.git
**/.bowerrc
**/bower_components
**/node_modules
**/npm-debug.log
```

Output is as expected with `docker build .`  

`docker-compose build --no-cache test` copied over everything

**Comment by shin-** [comment]
#2992 has some more `.dockerignore` fixes that should be included in 1.6.1

**Comment by stekershaw** [comment]
Thanks @shin- , unfortunately I still see different behaviour with docker-compose 1.6.2 from Docker Toolbox for Windows:

```
$ docker-compose --version
docker-compose version 1.6.2, build e80fc83

$ find .
.
./.dockerignore
./docker-compose.yml
./Dockerfile
./files
./files/test_dir
./files/test_dir/should_be_here_maybe
./files/test_dir/should_not_be_here

$ cat .dockerignore
files/test_dir
!files/test_dir/should_be_here_maybe

$ docker-compose build --no-cache
Building test
Step 1 : FROM busybox
latest: Pulling from library/busybox
f810322bba2c: Pull complete
a3ed95caeb02: Pull complete
Digest: sha256:97473e34e311e6c1b3f61f2a721d038d1e5eef17d98d1353a513007cf46ca6bd
Status: Downloaded newer image for busybox:latest
 ---> 3240943c9ea3
Step 2 : COPY . /context
 ---> 85a65d7f861c
Removing intermediate container 386d3103d8ab
Step 3 : CMD find /context
 ---> Running in e5e29b5746c4
 ---> 2c2d57a899ea
Removing intermediate container e5e29b5746c4
Successfully built 2c2d57a899ea

$ docker-compose run -d test
dcitest_test_run_1

$ docker logs dcitest_test_run_1
/context
/context/Dockerfile
/context/files
/context/files/test_dir
/context/files/test_dir/should_be_here_maybe
/context/files/test_dir/should_not_be_here
/context/.dockerignore
/context/docker-compose.yml
```

**Comment by shin-** [comment]
That's very odd. It definitely works for me.
```
$ docker-compose build --no-cache
Building web
Step 1 : FROM busybox
 ---> 3240943c9ea3
Step 2 : COPY . /context
 ---> 3619871879ad
Removing intermediate container 08432f688579
Step 3 : CMD find /context
 ---> Running in 5bbcf987c9e7
 ---> cf2bff2c1416
Removing intermediate container 5bbcf987c9e7
Successfully built cf2bff2c1416

$ docker-compose run -d web
Creating network "testdockerignore_default" with the default driver
testdockerignore_web_run_1

$ docker logs testdockerignore_web_run_1 
/context
/context/Dockerfile
/context/files
/context/files/test_dir
/context/files/test_dir/should_be_here
/context/docker-compose.yml
/context/.dockerignore

$ find .
.
./Dockerfile
./files
./files/test_dir
./files/test_dir/should_not_be_here
./files/test_dir/should_be_here
./docker-compose.yml
./.dockerignore

$ cat .dockerignore files/test_dir
!files/test_dir/should_be_here
```

What's the output of `pip show docker-py` in your environment?

**Comment by digizen** [comment]
I believe I also have the same issue. 

     Docker build .

works for me. However, enclosing the same build in a docker-compose.yml file breaks my build. In my .dockerignore file, I'm excluding the node_modules directory for my node build. This is the relevant section from my docker-compose.yml:
```
web:
  build: ./app
  ports:
    - "8080:8080"
  links:
    - mongodb
```

I believe I'm running the most recent version of docker-compose:
```
>docker-compose version
docker-compose version 1.6.2, build e80fc83
docker-py version: 1.7.2
CPython version: 2.7.11
OpenSSL version: OpenSSL 1.0.2d 9 Jul 2015

```

**Comment by shin-** [comment]
WIndows-specific issue, maybe with path separators?

**Comment by digizen** [comment]
Don't know how to tell whether this is related to path separators. 

Let me know if I can assist with testing or otherwise. Right now, I have a batch file that moves the "ignored" folders out of the project directory prior to the build, obviously ugly. 

Thanks for your help.

**Comment by digizen** [comment]
For completeness' sake, this is my .dockerignore file:
```
**/node_modules
```

**Comment by digizen** [comment]
Aha! It's the glob pattern that's causing the problem. removing the "**/" fixed the issue for me.

**Comment by digizen** [comment]
While I have a workaround for my specific issue, the fact remains that there is a bug. At the very least the glob pattern **/ doesn't get properly parsed with docker-compose, but works fine with the regular docker build command. Other glob patterns may or may not work.

**Comment by JonasT** [comment]
@bfirsh and everyone else who might care:

I have an absolutely simple .dockerignore which looks like this:
```
livedata
readonly-data
```

The .dockerignore doesn't work either (those two folders in the build context folder get uploaded to the daemon and it takes *forever*) and I'm on **Linux**. I updated just to current 1.6.2, no change. Unless I'm seeing a different problem with the same symptom, let me reiterate *again* that this doesn't appear to be windows-specific problem..

**Comment by nicbarker** [comment]
+1, also broken for me on Linux - don't think it's a windows specific problem.

**Comment by aanand** [comment]
@JonasT @nicbarker How are you confirming that the folders named in `.dockerignore` are being uploaded, other than the fact it's taking a long time? If you have a way of checking the tarball, I'd like to try and use it to reproduce the problem locally. As it is, I've added some debug to docker-py and am still unable to reproduce:

```console
$ pip list | grep docker
docker-compose (1.7.0.dev0, /Users/aanand/work/docker/compose)
docker-py (1.8.0rc2, /Users/aanand/work/docker/docker-py)

$ find .
.
./.dockerignore
./docker-compose.yml
./Dockerfile
./include.txt
./livedata
./livedata/exclude.txt
./readonly-data
./readonly-data/exclude.txt

$ cat .dockerignore
livedata
readonly-data

$ cat Dockerfile
FROM busybox
COPY . /data

$ docker-compose --verbose build --no-cache
<unrelated output>
docker.utils.utils.tar: Writing tar file to <open file '<fdopen>', mode 'w+b' at 0x1090f26f0>
docker.utils.utils.tar:   Adding .dockerignore
docker.utils.utils.tar:   Adding Dockerfile
docker.utils.utils.tar:   Adding docker-compose.yml
docker.utils.utils.tar:   Adding include.txt
docker.utils.utils.tar: Done
<unrelated output>
```

**Comment by JonasT** [comment]
@aanand it takes a REALLY long time before it starts, and when I move the Dockerfile to ./context and use "build: ./context/" with no other change it is instantly blazingly fast. Also back then when this problem first emitted I tried docker build manually with "docker build" and it was also instant.

**Comment by dnephin** [comment]
That doesn't really confirm the issue. If there were another large directory in the project root, moving the build to `./context` would make it faster.

**Comment by JonasT** [comment]
@dnephin there are no other folders in there except for "context" itself which I aded as a workaround, and three tiny text files..

**Comment by aanand** [comment]
@JonasT Is it possible you have a large `.git` directory that's getting included?

**Comment by JonasT** [comment]
No I checked with `ls -la`. There's really no other directory in there... or I have recently become blind :-)
```
jonas@cyberman:~/Server/nginx$ ls -la
total 56
drwxr-xr-x.  5 root  root 4096 Mar 30 05:05 .
drwxr-xr-x. 15 jonas root 4096 Mar 30 17:43 ..
drwxr-xr-x.  2 root  root 4096 Mar 30 05:05 context
-rw-r--r--.  1 root  root  796 Mar 30 05:05 docker-compose.yml
-rw-r--r--.  1 root  root   25 Feb 29 21:06 .dockerignore
drwxr-xr-x.  3  1011 root 4096 Feb 29 21:06 livedata
-rw-r--r--.  1 root  root   50 Feb 29 21:06 permissions.yml
drwxr-xr-x.  4  1011 root 4096 Feb 29 21:06 readonly-data
jonas@cyberman:~/Server/nginx$ ls -la context/
total 16
drwxr-xr-x. 2 root root 4096 Mar 30 05:05 .
drwxr-xr-x. 5 root root 4096 Mar 30 05:05 ..
-rw-r--r--. 1 root root  956 Mar 30 05:05 Dockerfile
jonas@cyberman:~/Server/nginx$ 
```
Both livedata and readonly-data are fairly huge (a couple of gigabytes).

**Comment by digizen** [comment]
Back to this being a Windows problem? :)

**Comment by ulrichSchreiner** [comment]
hi,
i have linux, compose 1.6.2, docker 1.10.3 ... exact the same problem. docker uses `.dockerignore`, compose ignores it.

**Comment by aanand** [comment]
@ulrichSchreiner Can you provide steps to reproduce?

**Comment by ulrichSchreiner** [comment]
hi,

here is a gist (https://gist.github.com/ulrichSchreiner/566815cea26ce55b95207e7795cf6962). the `.dockerignore` contains `**/node_modules` the Dockerfile adds a file in a `t1/a/b/c/node_modules` subfolder.

if you build this with "docker build ..." you get an error because there is no file to add (it is correctly ignored). you can see the output in the last file of the gist.

but if you build it with the given "docker-compose.yml" it is successfully build --> docker-compose ignores the `.dockerignore` file.

**Comment by aanand** [comment]
OK, I can reproduce this. Looks like docker-py's probably got a bug with `**/` rules.

**Comment by shin-** [comment]
IIRC, lack of support for the `**` syntax is a known limitation of our `.dockerignore` implementation. See https://github.com/docker/docker-py/pull/721#issuecomment-135065043

**Comment by mllanes** [comment]
Same issue here. `.dockerignore` is ignored when used with docker-compose
`docker-compose version 1.6.2, build 4d72027`

**Comment by lattwood** [comment]
```
./.dockerignore
./Dockerfile (symlink to ./server/docker/cms/Dockerfile)
./server/docker/docker-compose.yml
```

Contents of .dockerignore

```
./.git
./.vagrant
```

.vagrant is 16GB, so this is a "Big Deal(tm)" for us.

**Comment by lattwood** [comment]
Sorry, 
```
docker-machine version 0.7.0, build a650a40

docker-compose version 1.7.0, build 0d7bf73
docker-py version: 1.8.0
CPython version: 2.7.9
OpenSSL version: OpenSSL 1.0.1j 15 Oct 2014

Client:
 Version:      1.11.1
 API version:  1.23
 Go version:   go1.5.4
 Git commit:   5604cbe
 Built:        Tue Apr 26 23:44:17 2016
 OS/Arch:      darwin/amd64

Server:
 Version:      1.11.1
 API version:  1.23
 Go version:   go1.5.4
 Git commit:   5604cbe
 Built:        Wed Apr 27 00:34:20 2016
 OS/Arch:      linux/amd64
```

**Comment by aanand** [comment]
@lattwood Looks like a bug - I've fixed it in https://github.com/docker/docker-py/pull/1065. For now, you can remove the leading `./` from your patterns and it should work.

**Comment by lattwood** [comment]
Thanks!

**Comment by agilgur5** [comment]
Running on Ubuntu 14.04, my dockerignore is simply `**/always_restart.txt`. On a regular `docker build` (v1.11.2), I don't see the file in the container, but with Compose (v1.7.1), I do.

**Comment by aanand** [comment]
@agilgur5 Correct, this is a known issue with docker-py. I've created https://github.com/docker/docker-py/issues/1117 to track it.

**Comment by luckydonald** [comment]
My ugly workaround in the DOCKERFILE:
(Using [git-check-ignore](https://www.kernel.org/pub/software/scm/git/docs/git-check-ignore.html) for parsing, thus needing `git init`)
```dockerfile
RUN mkdir /app/
WORKDIR   /app/
RUN git init
COPY ./.dockerignore /app/
RUN mv .dockerignore .gitignore
RUN DEL_ME=$(find . -print0 | xargs -0) && echo Forcing deletion of .dockerignore files && git check-ignore --no-index $DEL_ME > tmp.txt
RUN DEL_ME=$(cat tmp.txt | xargs) && echo DELETING: $DEL_ME tmp.txt && rm -f -d -r $DEL_ME tmp.txt
```

**Comment by jameschao** [comment]
I think found a workaround for this by using data volumes in `docker-compose.yml`, by basically listing the container directories that you don't want to be overwritten. In our case, we wanted to preserve the `node_modules` directory in the container.

So, our `docker-compose.yml` looks like this:
```
  my_app:
    build: ./my_app
    volumes:
      - ./my_app:/app
      # prevent the mounting above from overwriting the following
      - /app/node_modules
```

**Comment by cjerdonek** [comment]
I have a high-level suggestion for this issue:

I would recommend retitling this issue to something like, "[meta issue]: incorrect handling of .dockerignore rules" (or "[tracking issue]: incorrect handling ..."), or else simply closing this issue in favor of a collection of smaller, more actionable issues. This is because (1) it's not that Docker compose is "ignoring" the `.dockerignore` file (which is what the issue title currently says). It simply hasn't implemented the rules correctly. And (2), there are a number of parts to this issue. This isn't just one issue. For example, there is [this one](https://github.com/docker/compose/issues/1607) (`**` processing) and [this new one](https://github.com/docker/compose/issues/3886) I just filed (last-line precedence). And there very well may be more.

Because this issue thread is so long (because it covers a broad problem area), the individual problems that need to be solved are hard to locate within the thread. As opposed to a general ".dockerignore doesn't seem to work" issue, I'm suggesting filing well-defined issues that can be solved in isolation. As currently phrased, this issue may never be closed because the Python implementation might never have parity with the reference implementation.

**Comment by Chaoste** [comment]
I've used `**/**/node_modules` but figured out that this one was not working. So I defined each node_modules directory in my subprojects to fix this bug.

Before this hotfix I had an "Access Right Error" when installing npm dependencies because they were already there (caused by copying with a not working dockerignore file).

**Comment by yaroslavyaroslav** [comment]
@aanand 
It seems that this is not `docker-compose build` bug, rather `docker-compose up` bug. In my setup I've have error that there is no file while `docker-compose build` if i add file which i'm copying in the Dockerfile `COPY` , so the build part goes correctly. But if i save that line in `.dockerignore` file and comment the line with the `COPY` and run `docker-compose up app` with the the `command: ls -la ./` line in `docker-composet.yml` file will be show.

Here is my `Dockerfile`:
```
FROM ruby:2.3
RUN apt-get update && \
    apt-get install -y \
            build-essential \
            nodejs && \
    apt-get clean && \
    rm -rf /var/lib/apt/lists/* /tmp/* /var/tmp/*
RUN mkdir -p /app
WORKDIR /app
COPY Gemfile Gemfile.lock ./
RUN gem install bundler && bundle install --jobs 20 --retry 5
CMD ["ls", "-la", "./"]
COPY . ./
EXPOSE 3000
```


Here's my project folder:

```
drwxr-xr-x  23 user  staff   782  1 ноя 13:46 .
drwxr-xr-x   6 user  staff   204  1 ноя 12:59 ..
-rw-r--r--@  1 user  staff  8196  1 ноя 14:33 .DS_Store
-rw-r--r--   1 user  staff    78  1 ноя 15:16 .dockerignore
-rw-r--r--   1 user  staff   543  1 ноя 14:26 .gitignore
-rw-r--r--@  1 user  staff  1699  1 ноя 15:16 Dockerfile
-rw-r--r--   1 user  staff  1727  1 ноя 14:33 Gemfile
-rw-r--r--   1 user  staff  4320  1 ноя 14:33 Gemfile.lock
-rw-r--r--   1 user  staff   374  1 ноя 12:58 README.md
-rw-r--r--   1 user  staff   227  1 ноя 12:58 Rakefile
drwxr-xr-x  10 user  staff   340  1 ноя 12:58 app
drwxr-xr-x   8 user  staff   272  1 ноя 12:58 bin
drwxr-xr-x  14 user  staff   476  1 ноя 12:58 config
-rw-r--r--   1 user  staff   130  1 ноя 12:58 config.ru
-rw-r--r--   1 user  staff   606  1 ноя 14:16 database.yml
drwxr-xr-x   4 user  staff   136  1 ноя 13:39 db
-rw-r--r--   1 user  staff   232  1 ноя 15:16 docker-compose.yml
drwxr-xr-x   4 user  staff   136  1 ноя 12:58 lib
drwxr-xr-x   4 user  staff   136  1 ноя 12:59 log
drwxr-xr-x   9 user  staff   306  1 ноя 12:58 public
drwxr-xr-x   9 user  staff   306  1 ноя 12:58 test
drwxr-xr-x   8 user  staff   272  1 ноя 12:59 tmp
drwxr-xr-x   3 user  staff   102  1 ноя 12:58 vendor
```

here's my `.dockerignore`:

```
Gemfile
.DS_Store
db/*.sqlite3
db/*.sqlite-journal
log/*
./tmp/
Dockerfile
README.rdoc
```

Here's my `docker-compose.yml`:
```
app:
  build: .

  command: ls -la ./
  
  volumes:
    - .:/app
  ports:
    - "3000:3000"
  links:
    - postgres
postgres:
  image: postgres:9.4
  ports:
    - "5432"
```

Here's output of the `docker build -t demo .`:

```
...
Step 8 : WORKDIR /app
 ---> Using cache
 ---> be78de5f53d3
Step 9 : COPY Gemfile Gemfile.lock ./
lstat Gemfile: no such file or directory
```

The same have `docker-compose build`:
```
Step 8 : WORKDIR /app
 ---> Using cache
 ---> be78de5f53d3
Step 9 : COPY Gemfile Gemfile.lock ./
ERROR: Service 'app' failed to build: lstat Gemfile: no such file or directory
```

So the error shows that the build works well with `.dockerignore` in both cases.

But not the `docker-compose up` one. In the `Dockerfile` above comment those lines:
 
```
COPY Gemfile Gemfile.lock ./
RUN gem install bundler && bundle install --jobs 20 --retry 5
```
Then both builds will ran correct and we're be able to run that containers.  Firstly `docker run demo`:
```
drwxr-xr-x 11 root root 4096 Nov  1 12:43 .
drwxr-xr-x 58 root root 4096 Nov  1 12:45 ..
-rw-r--r--  1 root root   86 Nov  1 12:35 .dockerignore
-rw-r--r--  1 root root  543 Nov  1 11:26 .gitignore
-rw-r--r--  1 root root 4320 Nov  1 11:33 Gemfile.lock
-rw-r--r--  1 root root  374 Nov  1 09:58 README.md
-rw-r--r--  1 root root  227 Nov  1 09:58 Rakefile
drwxr-xr-x 10 root root 4096 Nov  1 09:58 app
drwxr-xr-x  2 root root 4096 Nov  1 09:58 bin
drwxr-xr-x  5 root root 4096 Nov  1 09:58 config
-rw-r--r--  1 root root  130 Nov  1 09:58 config.ru
-rw-r--r--  1 root root  606 Nov  1 11:16 database.yml
drwxr-xr-x  2 root root 4096 Nov  1 10:39 db
-rw-r--r--  1 root root  232 Nov  1 12:16 docker-compose.yml
drwxr-xr-x  4 root root 4096 Nov  1 09:58 lib
drwxr-xr-x  2 root root 4096 Nov  1 09:59 log
drwxr-xr-x  2 root root 4096 Nov  1 09:58 public
drwxr-xr-x  8 root root 4096 Nov  1 09:58 test
drwxr-xr-x  3 root root 4096 Nov  1 09:58 vendor
```

Here is no `Gemfile` everything is correct. But not with the `docker-compose up app`:
```

app_1       | total 60
app_1       | drwxr-xr-x 23 root root  782 Nov  1 10:46 .
app_1       | drwxr-xr-x 58 root root 4096 Nov  1 12:47 ..
app_1       | -rw-r--r--  1 root root 8196 Nov  1 11:33 .DS_Store
app_1       | -rw-r--r--  1 root root   86 Nov  1 12:35 .dockerignore
app_1       | -rw-r--r--  1 root root  543 Nov  1 11:26 .gitignore
app_1       | -rw-r--r--  1 root root 1701 Nov  1 12:43 Dockerfile
app_1       | -rw-r--r--  1 root root 1727 Nov  1 11:33 Gemfile
app_1       | -rw-r--r--  1 root root 4320 Nov  1 11:33 Gemfile.lock
app_1       | -rw-r--r--  1 root root  374 Nov  1 09:58 README.md
app_1       | -rw-r--r--  1 root root  227 Nov  1 09:58 Rakefile
app_1       | drwxr-xr-x 10 root root  340 Nov  1 09:58 app
app_1       | drwxr-xr-x  8 root root  272 Nov  1 09:58 bin
app_1       | drwxr-xr-x 14 root root  476 Nov  1 09:58 config
app_1       | -rw-r--r--  1 root root  130 Nov  1 09:58 config.ru
app_1       | -rw-r--r--  1 root root  606 Nov  1 11:16 database.yml
app_1       | drwxr-xr-x  4 root root  136 Nov  1 10:39 db
app_1       | -rw-r--r--  1 root root  232 Nov  1 12:16 docker-compose.yml
app_1       | drwxr-xr-x  4 root root  136 Nov  1 09:58 lib
app_1       | drwxr-xr-x  4 root root  136 Nov  1 09:59 log
app_1       | drwxr-xr-x  9 root root  306 Nov  1 09:58 public
app_1       | drwxr-xr-x  9 root root  306 Nov  1 09:58 test
app_1       | drwxr-xr-x  8 root root  272 Nov  1 09:59 tmp
app_1       | drwxr-xr-x  3 root root  102 Nov  1 09:58 vendor
```
So the `Gemfile` is adds anyway. 

I'm not a pro in the Docker, but hope it could helps to figure the problem

**Comment by cerdiogenes** [comment]
I'm also having this problem in Ubuntu 16.04:
```shell
$ docker version                                                                                                                                                  22:23:06
Client:
 Version:      1.12.3
 API version:  1.24
 Go version:   go1.6.3
 Git commit:   6b644ec
 Built:        Wed Oct 26 22:01:48 2016
 OS/Arch:      linux/amd64

Server:
 Version:      1.12.3
 API version:  1.24
 Go version:   go1.6.3
 Git commit:   6b644ec
 Built:        Wed Oct 26 22:01:48 2016
 OS/Arch:      linux/amd64

$ docker-compose version
docker-compose version 1.8.1, build 878cff1
docker-py version: 1.10.3
CPython version: 2.7.9
OpenSSL version: OpenSSL 1.0.1e 11 Feb 2013
```

docker build works ok. Is there anything else I can do to help with this?

**Comment by podorozhny** [comment]
Same problem.

macOS Sierra 10.12.2
Docker version 1.12.5, build 7392c3b
docker-compose version 1.9.0, build 2585387

**Comment by shin-** [comment]
The `**` pattern issue should be fixed in 1.11.2
There are possible outstanding issues like the one with last-line precedence, which are tracked separately in #3931 and #3886. Please refer to those for updates.

**Comment by ekkis** [comment]
I'm not sure whether docker-compose is ignoring my `.dockerignore` or it's failing to interpret it correctly (unlike docker).  this is how I landed here: http://stackoverflow.com/questions/42883596/equivalent-builds-dont-behave-the-same

and here's my `.dockerignore` file (in case you can spot what the issue may be):

aprilis:Vertex ekkis$ cat .dockerignore 
```
*
!www
!app
!inc
*/node_modules
*/bower_components
**/*.log
**/Dockerfile
**/.gitignore
```

**Comment by aiomaster** [comment]
It seems that my `docker-compose build` ignores the `.dockerignore` file completely.
`docker build` works as expected.
```
$ docker-compose version                                                                                                                       
docker-compose version 1.14.0, build c7bdf9e
docker-py version: 2.3.0
CPython version: 2.7.13
OpenSSL version: OpenSSL 1.0.1t  3 May 2016
```
My `.dockerignore` looks like this
```
.bundle
.git
.gitignore
test
tmp
log
```
so nothing special here.
I am using Ubuntu 16.04 but I don't know if that matters.

BTW: Can someone explain, why `docker-compose build` does not use the same code as `docker build` to build an image using the `Dockerfile` and all that stuff? Why is there an obvious code duplication (that does not work really good as this issue shows)?

**Comment by ekkis** [comment]
why was this issue closed and how do we re-open it?  the software is clearly broken and needs to be fixed

**Comment by shin-** [comment]
https://github.com/docker/compose/issues/1607#issuecomment-287177186

**Comment by ekkis** [comment]
ah.  thanks shin.  glad it will get addressed

**Comment by aiomaster** [comment]
I've read the issues  #3931 and #3886 like suggested, but my .dockerignore is not special, so I could not map exactly my problem.
I doesn't work at all for me and it doesn't matter, whats in my .dockerignore.

**Comment by aiomaster** [comment]
So stupid, it was my mistake.:dizzy_face:
I mounted my code as a volume so it is clear, that there is nothing ignored. Sorry for that.
```
app:
  build: .
  container_name: my_app
  volumes:
    - .:/my_app
```

**Comment by MartinMuzatko** [comment]
I had the same problem. just when I noticed that I'm on an old compose version. Updated to `docker-compose version 1.22.0-rc1, build e7de1bc3`.

Now it works better, but still not the same like `docker build`

**Comment by shawnrushefsky** [comment]
This is still unresolved. Running Docker for Mac, latest version (Docker 18.03.1-ce-mac65). 

My .dockerignore:
```
**/package-lock.json
**/node_modules
```

but both are getting passed into the container when using docker-compose.

**Comment by shin-** [comment]
https://github.com/docker/docker-py/pull/2065 is in 1.22.0 RC2 and should address that.

**Comment by ascourtas** [comment]
I'm having the same issue. Linux, Docker version 18.03.1-ce

```
#.dockercompose
*.pyc
__pycache__
```

All .pyc and __pycache__ files are still present

**Comment by rodrigobdz** [comment]
I'm using Docker for Mac Version `18.06.1-ce-mac73 (26764)` and the bug is still there.

![image](https://user-images.githubusercontent.com/14152377/46458961-c140eb80-c7b7-11e8-8792-e6bbd6f11072.png)

**Comment by wladyslaws** [comment]
I'm using:

docker-compose version 1.22.0, build f46880f
Docker version 18.06.1-ce, build e68fc7a

running on Ubuntu 16.04 Xenial and docker-compose build ignores the following .dockerignore:

`**/*.jpg
**/*.png
**/*.pyc
**/*.solverstate
**/*.caffemodel
**/*.tgz
**/.pytest_cache
**/*__pycache__*
**/.git
**/node_modules
*.egg-info
.eggs
*Dockerfile*
build
dist
`
As it is supposed now to interpret double asterisks, I don't know what may be the cause. Anyone with the same issue?

**Comment by janhn** [comment]
> As it is supposed now to interpret double asterisks, I don't know what may be the cause. Anyone with the same issue?

I'm having the same issue (with ```**/.tox```).

**Comment by zymtx5g79k** [comment]
I had the same issue. I use different contexts.
**docker-compose.yml** is placed on app root path and looks like
```
services:
  api:
    build:
      context: ./docker/api
```
My files tree:
```
- app/
-- docker/
--- api/
---- .dockerignore <- it works!
-- docker-compose.yml
-- .dockerignore <- it does not work for contexts
```
Have a nice day!

**Comment by rodrigobdz** [comment]
@zymtx5g79k That's not working for me. 
My directory structure:
```
- .dockerignore
- docker/
-- development/
--- docker-compose.yml  
```

```
version: '3'
services:
  web:
    build:
      context: ../../.
```

**Comment by zymtx5g79k** [comment]
@rodrigobdz, check please:
**docker-compose.yml**:
```
services:
  api:
    build:
      context: ./../../
version: '3'
```
**.dockerignore**:
```
/docker
```
**Files structure**:
```
- docker/
-- api/
--- docker-compose.yml
- .dockerignore
```

**Comment by rodrigobdz** [comment]
Still the same outcome, the `.git` directory is in the container for example. I'll post here my .dockerignore. Maybe it has some issue itself.
```
# Custom
docker-compose.yml
docs/
livereload.js*
yarn-error.log
v8-compile-cache-0
# vim
*.swp

.git
.gitignore
README.md
```

**Comment by thaJeztah** [comment]
@rodrigobdz I'm not able to reproduce the issue (running docker compose version 1.22.0);

Prepare the test;

```bash
mkdir repro-1607 && cd repro-1607
mkdir -p docker/development/
mkdir -p ./.git/this-is-a-git-repo
echo "this is README.md" > README.md

cat > docker/development/docker-compose.yml -<<EOF
version: '3'
services:
  web:
    build:
      context: ../../.
EOF

cat > ./.dockerignore -<<EOF
# Custom
docker-compose.yml
docs/
livereload.js*
yarn-error.log
v8-compile-cache-0
# vim
*.swp

.git
.gitignore
README.md
EOF

cat > Dockerfile -<<EOF
FROM alpine
RUN apk add --no-cache tree
COPY . /foobar/
CMD tree -a /foobar/
EOF
```

Build using docker-compose;

```
cd docker/development/
docker-compose build --no-cache

Building web
Step 1/4 : FROM alpine
 ---> 11cd0b38bc3c
Step 2/4 : RUN apk add --no-cache tree
 ---> Running in 8767bc07dad9
fetch http://dl-cdn.alpinelinux.org/alpine/v3.8/main/x86_64/APKINDEX.tar.gz
fetch http://dl-cdn.alpinelinux.org/alpine/v3.8/community/x86_64/APKINDEX.tar.gz
(1/1) Installing tree (1.7.0-r1)
Executing busybox-1.28.4-r0.trigger
OK: 5 MiB in 14 packages
Removing intermediate container 8767bc07dad9
 ---> 3916d3b689bb
Step 3/4 : COPY . /foobar/
 ---> 76ab68d75f88
Step 4/4 : CMD tree -a /foobar/
 ---> Running in 9891624b3cab
Removing intermediate container 9891624b3cab
 ---> d22b81d149f2
Successfully built d22b81d149f2
Successfully tagged development_web:latest
```

And verify that the content is not added;

```bash
docker run --rm development_web:latest
/foobar/
├── Dockerfile
└── docker
    └── development
        └── docker-compose.yml

2 directories, 2 files
```

Now, rename the `.dockerignore` and perform the build again;


```bash
mv ../../.dockerignore ../../.dockerignore.disabled
docker-compose build --no-cache

Building web
Step 1/4 : FROM alpine
....
```

Verify that the content is no longer ignored, and added to the image:

```bash
docker run --rm development_web:latest
/foobar/
├── .dockerignore.disabled
├── .git
│   └── this-is-a-git-repo
├── Dockerfile
├── README.md
└── docker
    └── development
        └── docker-compose.yml

4 directories, 4 files
```

**Comment by rodrigobdz** [comment]
@thaJeztah Thanks for taking the time to look into it. I found the issue. 

My use case is similar to the one described in https://github.com/docker/compose/issues/2098#issue-108463351. I use a `COPY` command in my Dockerfile and additionally, mount the same directory as a **volume**. 

I was expecting the excludes in `.dockerignore` to be applied to the volume as described by https://github.com/docker/compose/issues/2098#issuecomment-143505943.

**Comment by thaJeztah** [comment]
> I was expecting the excludes in .dockerignore to be applied to the volume, as described in #2098 (comment).

No, when bind-mounting a path from the host, docker won't be "in the middle" of that; on Linux, it's literally mounting that directory from the host inside the container. On Docker Desktop (Docker for Mac / Windows), there's some additional "magic" involved to make those files available inside the VM where the daemon (and container) runs, but essentially it's doing the same as on Linux after that.

`.dockerignore` is only used during build, and was intended for speeding up builds; to prevent having to send files to the daemon that are not used/desired in the image.

**Comment by rodrigobdz** [comment]
Thanks for the clarification! I consider that this explanation should be in the docs to avoid further misunderstanding in the community. Currently, the [documentation](https://docs.docker.com/engine/reference/builder/#dockerignore-file) is more focused on how to ignore files rather than on the scope of `.dockerignore` itself.

**Comment by thaJeztah** [comment]
That page of the documentation is describing the Dockerfile and `docker build`; from that perspective; wondering if it would make sense to describe that it does not work for other commands/uses.

**Comment by rodrigobdz** [comment]
Well, if it is a common source of misunderstanding, and evidently it is, it would certainly make sense to have it mentioned there.

A **single sentence** in the documentation would have saved us @thaJeztah, and all the users below, **a lot of time**.

<details>
  <summary>Users misunderstanding the current documentation</summary>
<h2>Issue 1607</h2>

  ![1607](https://user-images.githubusercontent.com/14152377/47952026-43fcc800-df69-11e8-87c0-6a8a80238ce1.png)

<h2>Issue 2098</h2>

![2098](https://user-images.githubusercontent.com/14152377/47952028-465f2200-df69-11e8-9051-973a9b5de6e7.png)
</details>

**Comment by xvrqt** [comment]
Just wanted to +1 this issue.

I'm running docker-copose: 1.23.2 (newest I could find between Homebrew and PIP)
OS: Mac OS 10.14.2

It's still ignoring my .dockerfile while building. Going to have to copy things explicitly in the mean time.

**Comment by xvrqt** [comment]
> I fixed this issue by removing an unnecessary volume reference in my docker-compose file. Compose ignores .gitignore for volumes.

This fixed it for me - docker-compose will ignore a .dockerignore file in a mounted volume.

**Comment by Harvzor** [comment]
@xvrqt I've probably wasted hours over the last week having random errors with my config file not being properly copied (my build process is supposed to copy it but it was being blocked by the .dockerignore file not being used). This is kinda unbelievable.

Thanks for highlighting that.

**Comment by LocoDelAssembly** [comment]
Hi! What is the status of this? I've prepared very simple docker-compose.yml scenarios that do not involve volumes and yet it fails to pick up `.dockerignore` when building from remote git repo.

https://github.com/LocoDelAssembly/docker-compose-dockerignore

Is the error I'm having at master branch expected? Any solutions other than building and tagging the image with vanilla docker and then using it in docker-compose.yml?

Thanks

**Comment by thaJeztah** [comment]
> Any solutions other than building and tagging the image with vanilla docker and then using it in docker-compose.yml?

If you're running the current version of compose, you could use the `COMPOSE_DOCKER_CLI_BUILD=1` (and `DOCKER_BUILDKIT=1`) options to make docker compose use the native `docker build`.

Your example looks to be different than the one being discussed here, so might be good to open a new ticket (if there's none for that yet)

**Comment by LocoDelAssembly** [comment]
Thanks @thaJeztah! Works with `COMPOSE_DOCKER_CLI_BUILD=1`, but if I also add `DOCKER_BUILDKIT=1` it fails (for an unrelated reason).

https://travis-ci.org/LocoDelAssembly/docker-compose-dockerignore/builds/658351109

**Comment by thaJeztah** [comment]
Hm, I suspect that's a bug in the version of docker that's running there; docker 18.06 is pretty old (and EOL); that version of docker has quite an early version of BuildKit, which was not yet stable.

```
=> ERROR [internal] load metadata for docker.io/library/alpine:3.9.5      0.1s
218 => ERROR [1/5] FROM docker.io/library/alpine:3.9.5                        0.0s
219 => => resolve docker.io/library/alpine:3.9.5                              0.0s
220------
221 > [internal] load metadata for docker.io/library/alpine:3.9.5:
222------
223------
224 > [1/5] FROM docker.io/library/alpine:3.9.5:
225------
226failed to solve with frontend dockerfile.v0: failed to build LLB: failed to load cache key: docker.io/library/alpine:3.9.5 not found

**Comment by thaJeztah** [comment]
Oh, hm, I see your travis is installing docker as well (reading from my phone); does that replace the pre-installed version? If you add a `docker info` and `docker version` step after installing, does it show the correct (19.03.x) version of docker to be installed?

**Comment by LocoDelAssembly** [comment]
Here it is https://travis-ci.org/LocoDelAssembly/docker-compose-dockerignore/builds/658435194

# Summary
## Docker version for both failed builds
```
$ docker version
Client: Docker Engine - Community
 Version:           19.03.7
 API version:       1.40
 Go version:        go1.12.17
 Git commit:        7141c199a2
 Built:             Wed Mar  4 01:22:36 2020
 OS/Arch:           linux/amd64
 Experimental:      false
Server: Docker Engine - Community
 Engine:
  Version:          19.03.7
  API version:      1.40 (minimum version 1.12)
  Go version:       go1.12.17
  Git commit:       7141c199a2
  Built:            Wed Mar  4 01:21:08 2020
  OS/Arch:          linux/amd64
  Experimental:     false
 containerd:
  Version:          1.2.13
  GitCommit:        7ad184331fa3e55e52b890ea95e65ba581ae3429
 runc:
  Version:          1.0.0-rc10
  GitCommit:        dc9208a3303feef5b3839f4323d9beb36df0a9dd
 docker-init:
  Version:          0.18.0
  GitCommit:        fec3683
```
`docker info` output is a bit large so not including, but `Server Version: 19.03.7`.
## docker-compose versions
### DOCKER_COMPOSE_VERSION=1.26.0-rc2
```
$ docker-compose version
docker-compose version 1.26.0-rc1, build 07cab513
docker-py version: 4.2.0
CPython version: 3.7.6
OpenSSL version: OpenSSL 1.1.0l  10 Sep 2019
```
(Reported version is wrong, it is rc2, the build hash matches rc2)
###  DOCKER_COMPOSE_VERSION=1.25.4
```
$ docker-compose version
docker-compose version 1.25.4, build 8d51620a
docker-py version: 4.1.0
CPython version: 3.7.5
OpenSSL version: OpenSSL 1.1.0l  10 Sep 2019
```

**Comment by Joshua2504** [comment]
Issue still existant in 19.03.8.

**Comment by LocoDelAssembly** [comment]
> Issue still existant in 19.03.8.

Correct. Also updated my test repo with the newest 1.25 and 1.26 docker-compose versions but all continues to be as described in https://github.com/LocoDelAssembly/docker-compose-dockerignore

Build where the same kind of info I pasted above can be found: https://travis-ci.org/github/LocoDelAssembly/docker-compose-dockerignore/builds/673393656

**Comment by xpetit** [comment]
Have the same issue here, `.dockerignore` doesn't work in a reliable way with `docker-compose`

**Comment by nikkhn** [comment]
Also having the same issue, `.dockerignore` not being honored by `docker-compose` when building images. 
Docker version 19.03.13
Compose version 1.27.4

**Comment by costa** [comment]
Happy belated 5 years anniversary to this issue!

**Comment by Joshua2504** [comment]
😂

**Comment by MitulIslam** [comment]
Anyway, does anyone have a good work around for this?

**Comment by pedro2555** [comment]
Why is this closed?

**Comment by thaJeztah** [comment]
It was closed because the original issue was fixed through https://github.com/docker/compose/pull/2031, and https://github.com/docker/docker-py/pull/1065. After that, this ticket has become a bit of a kitchen-sink of anything (possibly) related to `.dockerignore`.

If you have an issue, suspect it's a bug, and have steps to reproduce it; please open a new ticket. For reproduction steps, make sure to describe steps as accurately as possible; if possible with a minimal code repository to reproduce the issue.

**Comment by pedro2555** [comment]
Ok, I get it. I'm not sure this is an issue with compose or docker. I've build a minimalist repo showing the issue. The second time you try to build you get the `error checking context: 'can't stat ... '`

[github.com/pedro2555/compose-issue-1607](https://github.com/pedro2555/compose-issue-1607)

```bash
git clone git@github.com:pedro2555/compose-issue-1607.git
cd compose-issue-1607/
docker-compose up --build
docker-compose up --build
```

**Comment by vasilvestre** [comment]
So we missed the 6th anniversary of the issue :D

**Comment by JRGit4UE** [comment]
The problem seems to be that we can't see / debug  / verify, what .dockerignore really does, so maybe some feedback, which line of .dockerignore has been 'activated' would create a faster understanding how things work, and save a lot of time..

**Comment by thaJeztah** [comment]
> The problem seems to be that we can't see / debug / verify, what .dockerignore really does, so maybe some feedback, which line of .dockerignore has been 'activated' would create a faster understanding how things work, and save a lot of time..

I think I wrote that as one of my "collected issues" on `.dockerignore` here; https://github.com/moby/moby/issues/40319 (feel free to participate on that ticket, or to describe your use-case)

**Comment by cmistiloglou** [comment]
Guys its almost the 10th anniversary of this issue ! Hope it gets fixed before 2025 arrives! 🎉

**Comment by thaJeztah** [comment]
@cmistiloglou https://github.com/docker/compose/issues/1607#issuecomment-830618429

> It was closed because the original issue was fixed through #2031, and [docker/docker-py#1065](https://github.com/docker/docker-py/pull/1065). After that, this ticket has become a bit of a kitchen-sink of anything (possibly) related to `.dockerignore`.
> 
> If you have an issue, suspect it's a bug, and have steps to reproduce it; please open a new ticket. For reproduction steps, make sure to describe steps as accurately as possible; if possible with a minimal code repository to reproduce the issue.
